### ISSUE TITLE
[JAVA] Unexpected end of stream exception generated while debugging with async POST with body

### ISSUE THREAD

**Issue by juanniosi** [swagger-api/swagger-codegen#4095]
##### Description

When doing any POST with any async method when APIClient.setDebugging(true)  we are getting an "java.net.ProtocolException: unexpected end of stream".
Setting APIClient.setDebugging(false) solves the issue, but we can't debug async POST calls.

Overriding onUploadProgress() I noticed that we get the double of bytesWritten than the contentLength.

With debugging in false

```
Upload progress bytesWritten=101, contentLength=101, done=true
Download progress bytesRead=51, contentLength=51, done=false
Download progress bytesRead=51, contentLength=51, done=true
```

With debugging in true

```
--> POST https://api.artik.cloud/v1.1/messages HTTP/1.1
Content-Type: application/json; charset=utf-8
Content-Length: 101
Authorization: Bearer <user-token-removed>
Accept: application/json
User-Agent: Swagger-Codegen/2.0.5/java
Upload progress bytesWritten=101, contentLength=101, done=true

{"data":{"volume":500},"sdid":"09174ee9e12644ec91f7ff4ed0e40700","ts":1477602273950,"type":"message"}
--> END POST (101-byte body)
Upload progress bytesWritten=202, contentLength=101, done=false
Exception in thread "OkHttp Dispatcher" java.lang.AssertionError: java.net.ProtocolException: unexpected end of stream
    at com.sun.tools.javac.util.Assert.error(Assert.java:133)
    at cloud.artik.api.MessagesApiTest$2.onFailure(MessagesApiTest.java:192)
    at cloud.artik.client.ApiClient$1.onFailure(ApiClient.java:1014)
    at com.squareup.okhttp.Call$AsyncCall.execute(Call.java:185)
    at com.squareup.okhttp.internal.NamedRunnable.run(NamedRunnable.java:33)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
```

I've tested this on swagger-codegen v2.2.2-SNAPSHOT.

**Comment by wing328** [comment]
@juanniosi thanks for reporting the issue. The default Java API client uses `okhttp-gson`. Can you try `retrofit2` or `jersey2` HTTP library instead to see if the issue persists?

You may find this useful: https://github.com/swagger-api/swagger-codegen/wiki/FAQ#how-can-i-generate-an-android-sdk

**Comment by mintu19** [comment]
I am facing the similar issue. When debug is on, 
java.net.ProtocolException: unexpected end of stream
error is produced at:
com.android.okhttp.internal.http.HttpConnection$FixedLengthSink.close()

**Comment by wing328** [comment]
@mintu19 have you tried other HTTP libraries (retrofit2, jersey2, etc) as mentioned above?

**Comment by mintu19** [comment]
@wing328 No I have not. As i was in hurry to provide workable code, so was working with okhttp only. Will check other libraries when i get time.

**Comment by cwacekINV** [comment]
I can confirm that this happens with retrofit2 as well. retrofit doesn't use the same debugging API, so that isn't enabled.

**Comment by wing328** [comment]
Here is a discussion related to the issue "unexpected end of stream"

https://github.com/square/okhttp/issues/2147#issuecomment-171768924

It mentions about 2 workarounds but I've not tried any of those. Please give it a try to see how it goes.

**Comment by Megamzero** [comment]
I found out  `HttpLoggingInterceptor` configured with `Level.BODY` can throw the error.

from the [okhttp HttpLoggingInterceptor repo](https://github.com/square/okhttp/blob/7135628c645892faf1a48a8cff464e0ed4ad88cb/okhttp-logging-interceptor/src/main/java/okhttp3/logging/HttpLoggingInterceptor.java#L188) the RequestBody is write out.
```
Buffer buffer = new Buffer();
requestBody.writeTo(buffer);
```
[right after](https://github.com/square/okhttp/blob/7135628c645892faf1a48a8cff464e0ed4ad88cb/okhttp-logging-interceptor/src/main/java/okhttp3/logging/HttpLoggingInterceptor.java#L212) it proceed the chain.

```
response = chain.proceed(request);
```
I think it's because the `RequestBody` was emptied out the `Unexpected end of stream` happen.

**Comment by Megamzero** [comment]
I just found what the cause of this error.
The Generated file `ProgressRequestBody` was only call in POST requests.
this class keep an instance of `BufferedSink` and use it each time `writeTo` method was called
it work fine with GET requests cause `writeTo` is call only one time.

I was able to correct the error by changing the `ProgressRequestBody.mustache` template.
removed line 26 :
```
    private BufferedSink bufferedSink;
````

modified writeTo method with :
```
    @Override
    public void writeTo(BufferedSink sink) throws IOException {
        BufferedSink bufferedSink = Okio.buffer(sink(sink));
        requestBody.writeTo(bufferedSink);
        bufferedSink.flush();

    }
```

**Comment by wing328** [comment]
@Megamzero may I know if you've time to submit a PR so that we can review the fix?

**Comment by talbright** [comment]
This is still an issue and a serious one, as it also affects GET requests. The fix suggested by @Megamzero works.

**Comment by wing328** [comment]
@talbright thanks for confirming the fix by @Megamzero works. Do you mind submitting a PR for the fix?

https://github.com/swagger-api/swagger-codegen/blob/master/modules/swagger-codegen/src/main/resources/Java/libraries/okhttp-gson/ProgressRequestBody.mustache

**Comment by wing328** [comment]
Closed via #5786
